File: android/app/src/main/java/io/govcopilot/portalworkspace/PortalWorkspacePlugin.kt
package io.govcopilot.portalworkspace

import android.content.Intent
import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin
import org.json.JSONObject

@CapacitorPlugin(name = "PortalWorkspace")
class PortalWorkspacePlugin : Plugin() {

    companion object {
        var instance: PortalWorkspacePlugin? = null
        var activeCall: PluginCall? = null
        var activeBundle: JSONObject? = null
        var activeAdapter: JSONObject? = null
        var sessionId: String? = null
    }

    override fun load() {
        super.load()
        instance = this
    }

    @PluginMethod
    fun open(call: PluginCall) {
        val bundleObj = call.getObject("bundle") ?: run {
            call.reject("bundle is required")
            return
        }

        activeCall = call
        activeBundle = JSONObject(bundleObj.toString())
        sessionId = "SES_${System.currentTimeMillis()}"

        val intent = Intent(context, PortalWorkspaceActivity::class.java).apply {
            putExtra(PortalWorkspaceActivity.EXTRA_BUNDLE_JSON, bundleObj.toString())
            addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
        }
        context.startActivity(intent)

        val ret = JSObject()
        ret.put("sessionId", sessionId)
        call.resolve(ret)
    }

    @PluginMethod
    fun close(call: PluginCall) {
        PortalWorkspaceActivity.instance?.finish()
        activeCall = null
        activeBundle = null
        activeAdapter = null
        sessionId = null
        call.resolve()
    }

    @PluginMethod
    fun fillCurrentPage(call: PluginCall) {
        val activity = PortalWorkspaceActivity.instance
        if (activity == null) {
            call.reject("Portal workspace is not open")
            return
        }
        activity.fillCurrentPage { result ->
            val ret = JSObject()
            ret.put("filled", result.filled)
            ret.put("failed", result.failed)
            ret.put("paused", result.paused)
            ret.put("step_id", result.stepId)
            ret.put("message", result.message)
            call.resolve(ret)
        }
    }

    fun emitEvent(type: String, caseId: String, message: String? = null, stepId: String? = null) {
        val payload = JSObject()
        payload.put("session_id", sessionId ?: "unknown")
        payload.put("case_id", caseId)
        payload.put("event_type", type)
        payload.put("timestamp", java.time.Instant.now().toString())
        if (message != null) payload.put("message", message)
        if (stepId != null) payload.put("step_id", stepId)
        notifyListeners("portalEvent", payload)
    }
}

File: native/portal-workspace-ios/PortalWorkspacePlugin.swift
import Foundation
import Capacitor

/// Capacitor bridge for Gov Copilot portal workspace (WKWebView + autofill).
@objc(PortalWorkspacePlugin)
public class PortalWorkspacePlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "PortalWorkspacePlugin"
    public let jsName = "PortalWorkspace"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "open", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "close", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "fillCurrentPage", returnType: CAPPluginReturnPromise),
    ]

    public static weak var shared: PortalWorkspacePlugin?
    public static var sessionId: String?
    public static var activeBundle: [String: Any]?
    public static var activeAdapter: [String: Any]?
    public static weak var workspace: PortalWorkspaceViewController?

    public override func load() {
        PortalWorkspacePlugin.shared = self
    }

    @objc func open(_ call: CAPPluginCall) {
        guard let bundle = call.getObject("bundle") else {
            call.reject("bundle is required")
            return
        }

        PortalWorkspacePlugin.activeBundle = bundle
        PortalWorkspacePlugin.sessionId = "SES_\(Int(Date().timeIntervalSince1970 * 1000))"

        DispatchQueue.main.async {
            guard let presenter = self.bridge?.viewController else {
                call.reject("No view controller")
                return
            }
            if PortalWorkspacePlugin.workspace != nil {
                call.reject("Portal workspace already open")
                return
            }

            let workspace = PortalWorkspaceViewController(plugin: self, bundle: bundle)
            workspace.modalPresentationStyle = .fullScreen
            presenter.present(workspace, animated: true)
            PortalWorkspacePlugin.workspace = workspace

            let ret = JSObject()
            ret["sessionId"] = PortalWorkspacePlugin.sessionId
            call.resolve(ret)
        }
    }

    @objc func close(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            PortalWorkspacePlugin.workspace?.dismissWorkspace()
            PortalWorkspacePlugin.workspace = nil
            PortalWorkspacePlugin.activeBundle = nil
            PortalWorkspacePlugin.activeAdapter = nil
            PortalWorkspacePlugin.sessionId = nil
            call.resolve()
        }
    }

    @objc func fillCurrentPage(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            guard let workspace = PortalWorkspacePlugin.workspace else {
                call.reject("Portal workspace is not open")
                return
            }
            workspace.fillCurrentPage { result in
                let ret = JSObject()
                ret["filled"] = result.filled
                ret["failed"] = result.failed
                ret["paused"] = result.paused
                if let stepId = result.stepId { ret["step_id"] = stepId }
                if let message = result.message { ret["message"] = message }
                call.resolve(ret)
            }
        }
    }

    func emitEvent(type: String, caseId: String, message: String? = nil, stepId: String? = nil) {
        let payload = JSObject()
        payload["session_id"] = PortalWorkspacePlugin.sessionId ?? "unknown"
        payload["case_id"] = caseId
        payload["event_type"] = type
        payload["timestamp"] = ISO8601DateFormatter().string(from: Date())
        if let message = message { payload["message"] = message }
        if let stepId = stepId { payload["step_id"] = stepId }
        notifyListeners("portalEvent", data: payload)
    }
}